lib/dao.js

```diff
@@ -509,6 +509,35 @@ DataAccessObject.upsert = function(data, options, cb) {
   if (id === undefined || id === null) {
     return this.create(data, options, cb);
   }
+  var doValidate = undefined;
+  if (options.validate === undefined) {
+    if (Model.settings.validateUpsert === undefined) {
+      if (Model.settings.automaticValidation !== undefined) {
+        doValidate = Model.settings.automaticValidation;
+      }
+    } else {
+      doValidate = Model.settings.validateUpsert;
+    }
+  } else {
+    doValidate = options.validate;
+  }
+
+  var forceId = this.settings.forceId;
+  if (forceId) {
+    options = Object.create(options);
+    options.validate = doValidate;
+    if (doValidate) {
+      Model.findById(id, function(err, model) {
+        if (err) return cb(err);
+        model.updateAttributes(data, options, cb);
+      });
+    } else {
+      const model = new Model({id: id}, {persisted: true});
+      options.validate = false;
```

Comment:
`options.validate = !!doValidate;` looks like the solution, this redundant is added for `options.validate === undefined`